libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -750,140 +983,154 @@ uint16_t GCS::sys_status_errors1()
-                                             : MAV_MISSION_ERROR) ;
+    // not a guided-mode reqest
+    MissionItemProtocol *prot = gcs().get_prot_for_mission_type(type);
+    if (prot == nullptr) {
+        send_mission_ack(*msg, type, MAV_MISSION_UNSUPPORTED);
+        return;
+    }
 
-        // verify we received the command
-        goto mission_ack;
+    if (!prot->receiving) {
+        send_mission_ack(*msg, type, MAV_MISSION_INVALID_SEQUENCE);
+        return;
     }
 
-    if (current == 3) {
-        //current = 3 is a flag to tell us this is a alt change only
-        // add home alt if needed
-        handle_change_alt_request(cmd);
+    prot->handle_mission_item(*msg, cmd);
+}
 
-        // verify we recevied the command
-        result = MAV_MISSION_ACCEPTED;
-        goto mission_ack;
+MAV_MISSION_RESULT MissionItemProtocol_Waypoints::replace_cmd(const AP_Mission::Mission_Command &cmd)
+{
+    if (!mission.replace_cmd(cmd.index, cmd)) {
+        return MAV_MISSION_ERROR;
     }
+    return MAV_MISSION_ACCEPTED;
+}
 
-    // Check if receiving waypoints (mission upload expected)
-    if (!waypoint_receiving) {
-        result = MAV_MISSION_ERROR;
-        goto mission_ack;
+MAV_MISSION_RESULT MissionItemProtocol_Waypoints::add_cmd(AP_Mission::Mission_Command &cmd)
+{
+    if (!mission.add_cmd(cmd)) {
+        return MAV_MISSION_ERROR;
+    }
+    return MAV_MISSION_ACCEPTED;
+}
+
+void MissionItemProtocol_Waypoints::complete(const GCS_MAVLINK &_link)
+{
+    _link.send_text(MAV_SEVERITY_INFO, "Flight plan received");
+    AP::logger().Write_EntireMission();
+}
+
+void MissionItemProtocol::handle_mission_item(const mavlink_message_t &msg, AP_Mission::Mission_Command &cmd)
+{
+    if (link == nullptr) {
+        AP::internalerror().error(AP_InternalError::error_t::gcs_bad_missionprotocol_link);
+        return;
     }
 
     // check if this is the requested waypoint
-    if (seq != waypoint_request_i) {
-        result = MAV_MISSION_INVALID_SEQUENCE;
-        goto mission_ack;
+    if (cmd.index != request_i) {
+        send_mission_ack(msg, MAV_MISSION_INVALID_SEQUENCE);
+        return;
     }
+    // make sure the item is coming from the system that initiated the upload
+    if (msg.sysid != dest_sysid) {
+        send_mission_ack(msg, MAV_MISSION_DENIED);
+        return;
+    }
+    if (msg.compid != dest_compid) {
+        send_mission_ack(msg, MAV_MISSION_DENIED);
+        return;
+    }
+
+    const uint16_t _item_count = item_count();
 
     // sanity check for DO_JUMP command
     if (cmd.id == MAV_CMD_DO_JUMP) {
-        if ((cmd.content.jump.target >= mission.num_commands() && cmd.content.jump.target >= waypoint_request_last) || cmd.content.jump.target == 0) {
-            result = MAV_MISSION_ERROR;
-            goto mission_ack;
+        if ((cmd.content.jump.target >= _item_count && cmd.content.jump.target > request_last) || cmd.content.jump.target == 0) {
```

Comment:
Pre exsisting error, but shouldn't the last part of this not be `cmd.content.jump.target == 0` but rather `cmd.content.jump.target < 1`

Reply:
Seems reasonable.